### PR TITLE
bpo-40563: Support pathlike objects on dbm/shelve

### PR DESCRIPTION
**Pull request by audeoudh**
Refers to https://bugs.python.org/issue40563.

This work continues @hakancelik96's work from #20274.

https://bugs.python.org/issue40563

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@audeoudh

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by gene1wood** [comment]
@audeoudh Will you be able to complete that CLA mentioned above so this PR can be used?

File: Lib/dbm/__init__.py [hunk 1]
```diff
@@ -109,6 +110,7 @@ def whichdb(filename):
     """
 
     # Check for ndbm first -- this has a .pag and a .dir file
+    filename = os.fspath(filename)
```
Comment (serhiy-storchaka) [inline Lib/dbm/__init__.py]: I some cases it is better to use `os.fsdecode()`. It will add support of bytes paths.

File: Lib/test/test_dbm.py [hunk 2]
```diff
@@ -129,6 +130,9 @@ def test_anydbm_access(self):
         assert(f[key] == b"Python:")
         f.close()
 
+    def test_open_with_pathlib_path(self):
+        dbm.open(Path(_fname), "c").close()
```
Comment (serhiy-storchaka) [inline Lib/test/test_dbm.py]: Use `os.support.os_helper.FakePath` for testing.

Test also with bytes path after adding support of it.

File: Lib/test/test_shelve.py [hunk 3]
```diff
@@ -63,29 +63,32 @@ def test_close(self):
         else:
             self.fail('Closed shelf should not find a key')
 
-    def test_ascii_file_shelf(self):
-        s = shelve.open(self.fn, protocol=0)
+    def test_open_template(self, filename=None, protocol=None):
+        kwargs = {
+            "filename": filename or self.fn,
+        }
+        if protocol:
```
Comment (serhiy-storchaka) [inline Lib/test/test_shelve.py]:
```suggestion
        if protocol is not None:
```

File: Modules/_dbmmodule.c [hunk 4]
```diff
@@ -470,10 +470,22 @@ dbmopen_impl(PyObject *module, PyObject *filename, const char *flags,
         return NULL;
     }
 
-    PyObject *filenamebytes = PyUnicode_EncodeFSDefault(filename);
-    if (filenamebytes == NULL) {
+    PyObject *filenamecharsorbytes = PyOS_FSPath(filename);
```
Comment (serhiy-storchaka) [inline Modules/_dbmmodule.c]: Use `PyUnicode_FSConverter()`.

File: Lib/dbm/__init__.py (hunk 1 above)
Comment (audeoudh) [reply Lib/dbm/__init__.py]: `os.fsencode()`, not `os.fsdecode()`, isn't it?  We have to take the object whatever it is (`str`, `bytes` or any object implementing `__fspath__`) and convert that to a `bytes` object.

Note that after `os.fsencode`, all extensions concatenated to that path should be bytes.

File: Modules/_dbmmodule.c (hunk 4 above)
Comment (audeoudh) [reply Modules/_dbmmodule.c]: Please review again, I'm not sure of `PyUnicode_FSConverter` usage.

**Comment by audeoudh** [comment]
It was a long time I did not check this PR.  I hope I did not forget something in the meantime.

I rebased the work on current main branch.  I took @serhiy-storchaka's review into account.

I suppose I should squash before the merge, shouldn't I?  Just let me know, I'll do it.

**Comment by erlend-aasland** [comment]
> I rebased the work on current main branch.

Force-pushing does not play well with GitHub and reviews. I would suggest to use `git merge --no-ff main` instead of `git rebase main`, as the review history is kept intact.

> I suppose I should squash before the merge, shouldn't I? Just let me know, I'll do it.

No need; all PR's are squashed upon merging :)

File: Lib/dbm/__init__.py [hunk 5]
```diff
@@ -75,6 +75,7 @@ def open(file, flag='r', mode=0o666):
             raise ImportError("no dbm clone found; tried %s" % _names)
 
     # guess the type of an existing database, if not creating a new one
+    file = os.fsencode(file)
```
Comment (serhiy-storchaka) [inline Lib/dbm/__init__.py]: Why was it added? Is it needed if `whichdb()` and `mod.open()` support path-like objects?

File: Lib/test/test_dbm.py [hunk 6]
```diff
@@ -129,6 +129,9 @@ def test_anydbm_access(self):
         assert(f[key] == b"Python:")
         f.close()
 
+    def test_open_with_pathlib_path(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_dbm.py:136]: Add please a test for bytes path (`os.fsencode(_fname)`).

File: Lib/test/test_dbm.py [hunk 7]
```diff
@@ -144,26 +147,27 @@ def setUp(self):
...
-            # Now add a key
-            f = module.open(_fname, 'w')
-            f[b"1"] = b"1"
-            # and test that we can find it
-            self.assertIn(b"1", f)
-            # and read it
-            self.assertEqual(f[b"1"], b"1")
-            f.close()
-            self.assertEqual(name, self.dbm.whichdb(_fname))
+        for path in [_fname, os_helper.FakePath(_fname)]:
```
Comment (serhiy-storchaka) [inline Lib/test/test_dbm.py]: Add please a bytes path.

File: Lib/test/test_dbm.py [hunk 8]
```diff
@@ -172,6 +176,10 @@ def test_whichdb_ndbm(self):
         with open(db_file, 'w'):
             self.addCleanup(os_helper.unlink, db_file)
         self.assertIsNone(self.dbm.whichdb(db_file[:-3]))
+        path = os_helper.FakePath(db_file)
+        with open(path, 'w'):
+            self.addCleanup(os_helper.unlink, path)
```
Comment (serhiy-storchaka) [inline Lib/test/test_dbm.py]: It is not needed. It creates an empty file and makes it removed after the test, but it is already done above.

File: Lib/test/test_dbm.py [hunk 9]
```diff
@@ -172,6 +176,10 @@ def test_whichdb_ndbm(self):
...
+        self.assertIsNone(self.dbm.whichdb(db_file[:-3]))
```
Comment (serhiy-storchaka) [inline Lib/test/test_dbm.py]: It is already tested at line 178. You need to test for `FakePath(db_file[:-3])` and `fsencode(db_file[:-3])`.

**Review by serhiy-storchaka — COMMENTED** [review]
The code LGTM. Please fix `test_whichdb_ndbm` and add tests for bytes paths.

File: Lib/test/test_dbm.py (hunk 6 above)
Comment (audeoudh) [reply Lib/test/test_dbm.py:136]: I now have tests for four cases:

- `str`
- `FakePath` with a `str` value
- `bytes`
- `FakePath` with a `bytes` value

AFAIK, all valid input should fall into one of these situations.

File: Lib/test/test_dbm.py (hunk 9 above)
Comment (audeoudh) [reply Lib/test/test_dbm.py]: OK, I misunderstood that test.  Should be OK now.

**Comment by audeoudh** [comment]
> Force-pushing does not play well with GitHub and reviews. I would suggest to use `git merge --no-ff main` instead of `git rebase main`, as the review history is kept intact.

True for GitHub reviews.  But `merge --no-ff` without `rebase` does not play well with `git log` itself.

But if all PR's are squased upon merging, it does not need to be rebased.  Thanks for the tip.

File: Doc/library/dbm.rst [hunk 10]
```diff
@@ -33,6 +33,8 @@ the Oracle Berkeley DB.
    file's format can't be guessed; or a string containing the required module
    name, such as ``'dbm.ndbm'`` or ``'dbm.gnu'``.
 
+.. versionchanged:: 3.10
+   Accepts :term:`path-like object` for filename.
```
Comment (audeoudh) [inline Doc/library/dbm.rst]: Hum, not sure of the Python version.  Will it be released in 3.10 or 3.11?

Comment (serhiy-storchaka) [reply Doc/library/dbm.rst]: 3.10 is a half step from release. Sorry, but it will be released in 3.11.

File: Lib/test/test_dbm_dumb.py [hunk 11]
```diff
@@ -294,6 +294,9 @@ def test_nonascii_filename(self):
             self.assertTrue(b'key' in db)
             self.assertEqual(db[b'key'], b'value')
 
+    def test_open_with_pathlib_path(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_dbm_dumb.py:297]: bytes?

File: Lib/test/test_dbm_gnu.py [hunk 12]
```diff
@@ -169,6 +169,9 @@ def test_nonexisting_file(self):
         self.assertIn(nonexisting_file, str(cm.exception))
         self.assertEqual(cm.exception.filename, nonexisting_file)
 
+    def test_open_with_pathlib_path(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_dbm_gnu.py:172]: bytes?

File: Lib/test/test_dbm_ndbm.py [hunk 13]
```diff
@@ -124,6 +124,9 @@ def test_nonexisting_file(self):
         self.assertIn(nonexisting_file, str(cm.exception))
         self.assertEqual(cm.exception.filename, nonexisting_file)
 
+    def test_open_with_pathlib_path(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_dbm_ndbm.py:127]: bytes?

File: Lib/test/test_shelve.py [hunk 14]
```diff
@@ -65,29 +65,26 @@ def test_close(self):
...
     def test_proto2_file_shelf(self):
-        s = shelve.open(self.fn, protocol=2)
-        try:
-            s['key1'] = (1,2,3,4)
-            self.assertEqual(s['key1'], (1,2,3,4))
-        finally:
-            s.close()
+        self.test_open_template(protocol=2)
+
+    def test_pathlib_path_file_shelf(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_shelve.py:87]: bytes?

File: Doc/library/dbm.rst (hunk 10 above)
Comment (audeoudh) [reply Doc/library/dbm.rst]: No problem for me, just have to fix the doc.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Pull request merged**